app/helpers/random_assign_helper.rb

```diff
@@ -0,0 +1,138 @@
+    # become an empty list when we've ensured that each reviewer has at least
+    # the amount of peer reviews we were told to assign.
+    while @eligible_reviewers.any?
+      @eligible_reviewers.each do |reviewer|
+        # Note: This function will perform the swap (if an index swap is needed),
+        # since we may have to look ahead and replace 'shuffle_index' with some
+        # element further ahead.
+        reviewee = get_next_reviewee_from_forward_search(reviewer, shuffle_index)
+
+        # If we get back nil, that means there was no possible swaps we could
+        # find from a forward search. Our last attempt is to look backwards to
+        # the previous elements, find whoever is assigned to them, and try to
+        # swap in the current element with one of their assignments whereby the
+        # trade would make this group and that one work. If we find none, then
+        # it's an error which should throw an exception.
+        if reviewee.nil?
+          assign_reviewee_backwards_or_throw(reviewer, shuffle_index)
+        else
+          add_peer_review_to_db_and_remember_assignment(reviewer, reviewee)
+        end
+
+        shuffle_index += 1
+      end
+
+      # Now that we can safely prune the list of groups that meet the required
+      # number of peer reviews, and then start again if we have no pruned out
+      # all the groups.
+      remove_ineligible_reviewers()
+    end
+  end
+
+  def remove_ineligible_reviewers
+    @eligible_reviewers.delete_if do |reviewer|
+      @reviewer_ids_assigned_reviewee_ids_map[reviewer.id].size >= @num_groups_for_reviewers
+    end
+  end
+
+  def get_next_reviewee_from_forward_search(reviewer, shuffle_index)
+    # First, check if the current index provided works.
+    potential_reviewee = @shuffled_reviewees[shuffle_index]
+    if eligible_to_be_assigned(reviewer, potential_reviewee)
+      return potential_reviewee
+    end
+
+    # Since the index doesnt work, look at the next element until the end of
+    # the list, and swap if we find a working group.
+    ((shuffle_index + 1)...@shuffled_reviewees.size).each do |new_shuffle_index|
+      potential_reviewee = @shuffled_reviewees[new_shuffle_index]
+      if eligible_to_be_assigned(reviewer, potential_reviewee)
+        swap_shuffled_indices(shuffle_index, new_shuffle_index)
+        return potential_reviewee
+      end
+    end
+
+    # Return nil if forward searching failed, which indicates that we will have
+    # to do backward searching.
+    nil
+  end
+
+  def eligible_to_be_assigned(reviewer, reviewee)
+    # If they already have an assignment to the group, it's not viable.
+    if @reviewer_ids_assigned_reviewee_ids_map.has_key?(reviewer.id) and
+          @reviewer_ids_assigned_reviewee_ids_map[reviewer.id].member?(reviewee.id)
+      return false
+    end
+
+    # Since they're not assigned, the returned boolean is if they don't share students.
+    reviewer.does_not_share_any_students?(reviewee)
+  end
```

Comment:
The above has been shortened (I guess git only shows the diff of the code at the time of the comment, rather than what it currently is... which is nicer).